### ISSUE TITLE
Some issues on Big Endian system

### ISSUE THREAD

**Issue by Link4Electronics** [afritz1/OpenTESArena#265]
Hello, I hope everything is fine. I compiled the project on a PowerMac G5 running Debian sid ppc64 and using a AMD Radeon R5 230 Terascale 2 gpu, The game works but has some issues as the screenshot attached. I understand that this can have a very low priority since not everyone has a big endian system laying around, but still would like to share the findings. Ingame it doesn't render the walls, it' an infinite map, and there's some font issues. Another thing is on powerpc there's no -march=native either as issue #248, it accepts -mcpu=native or even better -maltivec for G4 and up powerpc cpus.

![Screenshot_2024-11-24_06-27-05](https://github.com/user-attachments/assets/53edecbe-9e2e-4852-888c-23bfaba155ec)
![Screenshot_2024-11-24_06-27-36](https://github.com/user-attachments/assets/218eb109-ddb4-4d39-98b4-f3ae898b5609)
![Screenshot_2024-11-24_06-27-55](https://github.com/user-attachments/assets/f50a01a2-849b-4f80-b04b-35600e45c24c)

Best regards,
Link.

**Comment by afritz1** [comment]
It's shocking to see big endian. Let me understand first - it's a 20 y/o machine with recent GCC?

I'd like to fix these problems as there's already been effort to be careful with endianness. I can guess where the problems are based on these screenshots, but please post more if there are other cases besides font rendering and walls.

**Comment by Link4Electronics** [comment]
Yeap, late 2005 powermac g5 model (pci-express). running linux kernel 6.11.10, gcc 14, mesa drivers expose opengl 3.2 for  amd gpu and can do vulkan via software. I compiled a lot of opensource projects, noticed that sometimes the issue only affects PPC64, PPC32 works fine. just out of curiosity, GZdoom has a very similar issue, it doesn't render the wall textures and is just an empty room.

The intro, music and sound effects plays fine
![Screenshot_2024-11-26_01-10-00](https://github.com/user-attachments/assets/8b01f255-b76d-4960-b061-d860681ea836)
![Screenshot_2024-11-26_01-10-53](https://github.com/user-attachments/assets/3d7e74e7-eb8d-4235-9b1b-9a7133675d2e)
![Screenshot_2024-11-26_01-13-47](https://github.com/user-attachments/assets/30bff51a-d855-4453-a250-a6571cb87175)
![Screenshot_2024-11-26_01-14-27](https://github.com/user-attachments/assets/cedec7da-b420-4f89-bd47-b7d16f85e59f)

Best regards,
Link.

**Comment by afritz1** [comment]
Fix for font width in https://github.com/afritz1/OpenTESArena/commit/c8e1b9338dfb69260524068cda44cbd28c36d94f

Looking at map voxels, not immediately obvious why that's broken since the engine is reading the original map data in little endian. Maybe these places? I'm rusty at this.

https://github.com/afritz1/OpenTESArena/blob/c8e1b9338dfb69260524068cda44cbd28c36d94f/OpenTESArena/src/World/MapGeneration.cpp#L462
https://github.com/afritz1/OpenTESArena/blob/c8e1b9338dfb69260524068cda44cbd28c36d94f/OpenTESArena/src/World/MapGeneration.cpp#L1680
https://github.com/afritz1/OpenTESArena/blob/c8e1b9338dfb69260524068cda44cbd28c36d94f/OpenTESArena/src/World/ArenaInteriorUtils.cpp#L24

**Comment by Link4Electronics** [comment]
For the last part I thought about

#ifdef ENDIAN_BIG { //something like that
	return voxel | (voxel << 8);
} else {
	return (voxel << 8) | voxel;
}

**Comment by afritz1** [comment]
I ruled out `Compression::decodeType08()` because images look right.

I'm still stumped, so what I would try next if I were you is `printf()` some integers like `map1Voxel` on that computer and a modern computer and see if the values are different. The results from `Bytes::getLE16()` should be the same between computers.

**Comment by Link4Electronics** [comment]
QEMU works, but is kinda slow, it emulates a 900MHz powerpc,  I'm having a hard time to remember what I did to Jolt Physics to make it compile again. There's the instruction https://wiki.qemu.org/Documentation/Platforms/PowerPC, you wanna go with qemu-system-ppc64 -M mac99,via=pmu -m 2G -cpu 970 -hda <hd image file> -cdrom <iso file of installation media> you can use Debian sid ppc64 https://cdimage.debian.org/cdimage/ports/12.0/ppc64/ (you wanna go with this version to boot) or adelie linux ppc64 https://www.adelielinux.org/download/ but I don't know, I think it's gonna be too much effort, it's gonna take a while for me to test and compile. Other route and better solution would be cross compile for ppc64 with msys2 on windows and a gnu toolchain targeting ppc64.

Best regards,
Link.

**Comment by Link4Electronics** [comment]
Compiling rn, Jolt finishes built target.
35% we'll see.

**Comment by Link4Electronics** [comment]
Managed to compile, but now I get on console
[Assets/ExeData.cpp(958)] Error: Couldn't get "AttributeNames" (section "Entities").

Best regards,
Link.

**Comment by afritz1** [comment]
Re-copy the data and options folders over the ones in your build directory to update those ExeData strings.

**Comment by Link4Electronics** [comment]
Kudos!
![Screenshot_2024-11-27_01-57-48](https://github.com/user-attachments/assets/dcd59f60-113c-4281-b9ca-655ece4d76fd)
![Screenshot_2024-11-27_01-58-04](https://github.com/user-attachments/assets/4bc5ecd0-c108-4863-8383-d50ec4286f4f)
![Screenshot_2024-11-27_01-59-55](https://github.com/user-attachments/assets/f6220f37-f230-4b9c-8165-85c89d6b8a81)

**Comment by Link4Electronics** [comment]
Running with the console, got this warnings:

```
[World/MapGeneration.cpp(516)] Warning: Missing *CEILING texture ID for raised platform floor.
[Assets/INFFile.cpp(987)] Warning: Invalid *BOXSIDE index "13".
[World/MapGeneration.cpp(516)] Warning: Missing *CEILING texture ID for raised platform floor.
[World/MapGeneration.cpp(516)] Warning: Missing *CEILING texture ID for raised platform floor.
[World/MapGeneration.cpp(516)] Warning: Missing *CEILING texture ID for raised platform floor.
[World/MapGeneration.cpp(516)] Warning: Missing *CEILING texture ID for raised platform floor.
[Voxels/ArenaVoxelUtils.cpp(164)] Error: Couldn't get .INF voxel texture filename for ID "-1".
[Voxels/ArenaVoxelUtils.cpp(178)] Error: Couldn't get .INF voxel texture set index for ID "-1".
[World/MapGeneration.cpp(516)] Warning: Missing *CEILING texture ID for raised platform floor.
[Assets/INFFile.cpp(987)] Warning: Invalid *BOXSIDE index "15".
[World/MapGeneration.cpp(516)] Warning: Missing *CEILING texture ID for raised platform floor.
[Player/Player.cpp(109)] Error:
Need characterVirtual's contact listeners
```
We'll see if a can printf() map1Voxel
Best regards,
Link.

**Comment by afritz1** [comment]
Those are validation errors for the original game's data, Arena is just buggy in general.

I need to finish CharacterVirtual work sometime. Right now the Jolt character controller usage is the bare minimum to move the player around to test.

**Comment by Link4Electronics** [comment]
Here's some printf of map1Voxel, I'll edit for big endian asap

```
	{
		DebugAssert(map1Voxel != 0);
		DebugAssert(mostSigNibble != 0x8);

		printf("First printf: %u\n", (unsigned long)map1Voxel);

		if ((map1Voxel & 0x8000) == 0)
		{
			const uint8_t mostSigByte = ArenaLevelUtils::getVoxelMostSigByte(map1Voxel);
			const uint8_t leastSigByte = ArenaLevelUtils::getVoxelLeastSigByte(map1Voxel);
			const bool voxelIsSolid = mostSigByte == leastSigByte;
			printf("Second printf: %u\n", (unsigned long)map1Voxel);

			if (voxelIsSolid)
			{
				// Regular solid wall.
				*outVoxelType = ArenaTypes::VoxelType::Wall;
				outShapeInitCache->initDefaultBoxValues();
				ArenaMeshUtils::WriteWallRendererGeometryBuffers(outShapeInitCache->verticesView, outShapeInitCache->normalsView, outShapeInitCache->texCoordsView);
				ArenaMeshUtils::WriteWallRendererIndexBuffers(outShapeInitCache->opaqueIndices0View, outShapeInitCache->opaqueIndices1View, outShapeInitCache->opaqueIndices2View);

				const int textureIndex = mostSigByte - 1;
				const int clampedTextureID = ArenaVoxelUtils::clampVoxelTextureID(textureIndex);
				*outTextureAsset0 = TextureAsset(
					ArenaVoxelUtils::getVoxelTextureFilename(clampedTextureID, inf),
					ArenaVoxelUtils::getVoxelTextureSetIndex(clampedTextureID, inf));
				*outTextureAsset1 = *outTextureAsset0;
				*outTextureAsset2 = *outTextureAsset0;
				printf("Third printf: %u\n", (unsigned long)map1Voxel);
			}
```

Little endian system:
First printf: 7196
Second printf: 7196
Third printf: 7196
First printf: 7453
Second printf: 7453
Third printf: 7453
First printf: 6425
Second printf: 6425
Third printf: 6425
First printf: 6939
Second printf: 6939
Third printf: 6939
First printf: 6682
Second printf: 6682
Third printf: 6682
First printf: 6144
Second printf: 6144
[World/MapGeneration.cpp(518)] Warning: Missing *CEILING texture ID for raised platform floor.
First printf: 45211
First printf: 45213
First printf: 45072
First printf: 6152
Second printf: 6152
[World/MapGeneration.cpp(518)] Warning: Missing *CEILING texture ID for raised platform floor.
First printf: 3855
Second printf: 3855
Third printf: 3855

Big endian:

**Comment by Link4Electronics** [comment]
Hello again, I hope you had a good thanksgiving day.

I tested again and remembered that in the main menu of the game there is a Test mode, I tested some scenarios and noticed that only the Dungeon generates itself and with the map.
![Screenshot_2024-11-29_02-39-47](https://github.com/user-attachments/assets/84c14aab-569e-45ab-a888-d3af54da8eb3)
![Screenshot_2024-11-29_02-33-03](https://github.com/user-attachments/assets/db027163-4dab-4673-8a72-c72a5e5569b7)
![Screenshot_2024-11-29_02-34-21](https://github.com/user-attachments/assets/bf42af1c-1b2d-4eeb-9997-08b4733b1cc1)
 The other scenarios like main quest, village do not generate the maps, it is like in the prints above from the other days that I sent. In the village test it renders some NPCs but the world is empty.
![Screenshot_2024-11-29_02-39-18](https://github.com/user-attachments/assets/a056876c-b0d8-4ec2-b6ef-aa90758ebc27)

Best regards,
Link.

**Comment by afritz1** [comment]
I'm not going to get to this for a while, others are welcome to try. I still think it's how Arena map data is being decoded

**Comment by Link4Electronics** [comment]
Here's a qemu image: https://archive.org/details/otesa-ppc64
Extract it and run with
`qemu-system-ppc64 -M mac99,via=pmu -cpu 970fx -m 2G -hda otesa.img -boot c`
It takes a while to boot specially when initializing ramdisk, and it's very slow on qemu (need some patience) but at least it works.

user: aaron
pass: asd123
user: root
pass: asd123
user aaron belongs to the sudoers list

There's kate, fastfetch, libsdl2-dev, libsdl3-dev, OpenTESArena dependencies and an IDE KDevelop if needed. The source of OpenTESArena and otesacompiled are located at ~/home/aaron/Downloads and there's a Jolt Physics patch for powerpc if needed.

I forgot to disable compositing, I suggest disabling it as gonna help with performance a little bit, Applications>Settings>Windows Manager Tweaker>Compositor uncheck enable display compositing

There's no audio btw, I don't know how to make it work in qemu. If you have any problems with qemu and/or image, I'll be around to try to help.

Best regards,
Link.

**Comment by Link4Electronics** [comment]
Jolt Physics added support for [PowerPC](https://github.com/jrouwe/JoltPhysics/commit/9b3e9e604145eee5845edf54fb1fc05585fa61c0) (unfortunately only little endian for now, but it's not required to patch it anymore to compile)

Best regards,
Link.

**Comment by afritz1** [comment]
Did more fixes today, here's the most relevant one: 50023cf46a9773c812184ffc44de62204b0aada9
If this isn't it then I'm super stuck.

**Comment by Link4Electronics** [comment]
Yeap, that did the trick! Congratulations!
I made sure to test other scenarios to see if it was generating them and seems to be alright.
![Screenshot_2024-12-22_02-36-54](https://github.com/user-attachments/assets/b20861e1-30eb-45d6-b60b-6ac09ea33dab)
![Screenshot_2024-12-22_02-37-26](https://github.com/user-attachments/assets/97241869-d61c-4570-a3eb-2535f0a42432)
![Screenshot_2024-12-22_02-58-02](https://github.com/user-attachments/assets/71d6d7c5-5afb-4734-ae90-a2818f14d4d0)
Guess this issue can finally be closed.

Best regards,
Link.
